src/server/visibility/client_visibility.rs

```diff
@@ -0,0 +1,109 @@
+use bevy::{
+    ecs::entity::{EntityHashMap, EntityHashSet},
+    platform::collections::hash_map::Entry,
+    prelude::*,
+};
+
+/// Cached visibility information for a client.
+///
+/// Automatically updated by observers from [`AppVisibilityExt`](super::AppVisibilityExt).
+#[derive(Component, Default)]
+pub struct ClientVisibility {
+    /// List of hidden entities and the filters that block their visibility.
+    ///
+    /// Dilters are stored as a bitmask, with bit indices assigned by the
+    /// [`FilterRegistry`](super::registry::FilterRegistry).
+    hidden: EntityHashMap<u32>,
```

Comment:
32 filters seems very limiting. I recommend using feature flags to allow larger bit counts (64, 128, and an option for up to 2^16 filters that uses a `SmallVec<[u64; 2]>`; the `filter_bit` variable would need to be u16 instead of u8). The default can be 32 filters if no feature flag is set.